### DISCUSSION TITLE
java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "seq" is null

### DISCUSSION THREAD

**Discussion by SpigotRCE** [bivashy/NanoLimboPlugin#19] [General]
java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "seq" is null
        at io.netty.buffer.ByteBufUtil.utf8Bytes(ByteBufUtil.java:1103) ~[velocity.jar:3.4.0-SNAPSHOT (git-99aaf3ce-b439)]
        at ua.nanit.limbo.protocol.ByteMessage.writeString(ByteMessage.java:98) ~[?:?]
        at ua.nanit.limbo.protocol.packets.login.PacketLoginSuccess.encode(PacketLoginSuccess.java:48) ~[?:?]
        at ua.nanit.limbo.protocol.PacketSnapshot.encode(PacketSnapshot.java:50) ~[?:?]
        at ua.nanit.limbo.protocol.PacketSnapshot.of(PacketSnapshot.java:85) ~[?:?]
        at ua.nanit.limbo.connection.PacketSnapshots.<init>(PacketSnapshots.java:119) ~[?:?]
        at ua.nanit.limbo.server.LimboServer.start(LimboServer.java:96) ~[?:?]
        at org.nexuscraft.corevelocity.misc.LimboManager.createLimboServer(LimboManager.java:55) ~[?:?]
        at org.nexuscraft.corevelocity.misc.ServerHandler.onLimboRequest(ServerHandler.java:42) ~[?:?]
        at org.nexuscraft.corevelocity.listener.PlayerServerListener.onConnectionRequest(PlayerServerListener.java:51) ~[?:?]
        at org.nexuscraft.corevelocity.listener.Lmbda$42.execute(Unknown Source) ~[?:?]
        at com.velocitypowered.proxy.event.UntargetedEventHandler$VoidHandler.lambda$buildHandler$0(UntargetedEventHandler.java:56) ~[velocity.jar:3.4.0-SNAPSHOT (git-99aaf3ce-b439)]
        at com.velocitypowered.proxy.event.VelocityEventManager.fire(VelocityEventManager.java:676) ~[velocity.jar:3.4.0-SNAPSHOT (git-99aaf3ce-b439)]
        at com.velocitypowered.proxy.event.VelocityEventManager.lambda$fire$5(VelocityEventManager.java:541) ~[velocity.jar:3.4.0-SNAPSHOT (git-99aaf3ce-b439)]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]

**Comment by Zffu** [comment]
what caused this error?

**Comment by bivashy** [comment]
1. Client version?
2. Is client modded (Fabric/Forge/NeoForge or anything else)
3. ViaVersion on proxy?
4. Steps to reproduce

**Reply by SpigotRCE** [reply]
1.21.1
Fabric
no
idk

when using the config it works fine but with API this happens

**Reply by bivashy** [reply]
Try with vanilla Minecraft, Fabric could be the cause of the bug.

**Reply by SpigotRCE** [reply]
I don't think so, I tried with vanilla too but didn't work, when creating a limbo from config, it works fine for both vanilla and fabric but not with the API

**Comment by SpigotRCE** [comment]
That's what I am asking

**Comment by bivashy** [comment]
Your config `#getPingData()#getVersion()` returns `null`.

It tries to write "username" into `PacketLoginSuccess`, see: https://github.com/bivashy/NanoLimboPlugin/blob/ed238e7483d61db33ce8ef22ffae1488348ed36e/api/src/main/java/ua/nanit/limbo/protocol/packets/login/PacketLoginSuccess.java#L48

It writes null, but it should be declared here: https://github.com/bivashy/NanoLimboPlugin/blob/ed238e7483d61db33ce8ef22ffae1488348ed36e/api/src/main/java/ua/nanit/limbo/connection/PacketSnapshots.java#L91

**Comment by bivashy** [comment]
Solution: Config should declare `username` instead of simply returning `new PingData()`

**Comment by SpigotRCE** [comment]
Ok it worked, but what do I set in protocol version?

**Reply by bivashy** [reply]
> Ok it worked, but what do I set in protocol version?

https://wiki.vg/Protocol_version_numbers

![protocol-versions](https://github.com/user-attachments/assets/606b2aa3-6450-4faa-be75-51727778fdec)

**Reply by SpigotRCE** [reply]
I know that but how does it sets the protocol version if the limbo is created from the config?

**Reply by bivashy** [reply]
> I know that but how does it sets the protocol version if the limbo is created from the config?

https://github.com/bivashy/NanoLimboPlugin/blob/main/api/src/main/java/ua/nanit/limbo/protocol/packets/status/PacketStatusResponse.java

**Reply by SpigotRCE** [reply]
I have seen that but idk why it didn't work for me, I even tried to set the protocol version to the Player's protocol version but didn't work

**Reply by SpigotRCE** [reply]
any update?

**Reply by bivashy** [reply]
> any update?

i don't have idea why it doesn't work, but you could try ask nan1t, or elsewhere

**Reply by SpigotRCE** [reply]
I legit don't know too, but if I do the same thing but with a config file, it works fine, maybe we can hop on discord and try fix this?

**Comment by SpigotRCE** [comment]
`package org.nexuscraft.corevelocity.misc;

import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.proxy.server.ServerInfo;
import org.nexuscraft.corevelocity.Core;
import org.slf4j.Logger;
import ua.nanit.limbo.configuration.LimboConfig;
import ua.nanit.limbo.server.Command;
import ua.nanit.limbo.server.CommandHandler;
import ua.nanit.limbo.server.LimboServer;
import ua.nanit.limbo.server.data.BossBar;
import ua.nanit.limbo.server.data.InfoForwarding;
import ua.nanit.limbo.server.data.PingData;
import ua.nanit.limbo.server.data.Title;

import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Random;

public class LimboManager {
    private final ProxyServer proxyServer;
    private final Core plugin;
    private final Logger logger;
    private final HashMap<LimboServer, Integer> limboServers; // Limbo Instance, port number
    private final Random random;

    public LimboManager(ProxyServer proxyServer, Core plugin, Logger logger) {
        this.proxyServer = proxyServer;
        this.plugin = plugin;
        this.logger = logger;
        this.limboServers = new HashMap<>();
        this.random = new Random();

        logger.info("Started LimboManager");
    }

    public HashMap<LimboServer, Integer> getLimboServers() {
        return limboServers;
    }

    public LimboServer createLimboServer() {
        int port = 0;
        while (port == 0) {
            port = random.nextInt(65535) + 1024;
            if (!limboServers.containsValue(port) && port != 25565 && port != 19132) {
                break;
            }
        }

        LimboServer limboServer = new LimboServer(new Config("local-limbo-" + port, port), new CommandHandler_(), getClass().getClassLoader());
        limboServers.put(limboServer, port);
        try {
            limboServer.start();
        } catch (Exception e) {
            logger.error("Failed to start Limbo Server", e);
            return null;
        }
        proxyServer.registerServer(
                new ServerInfo("local-limbo-" + port,
                        (InetSocketAddress) limboServer.getConfig().getAddress()
                )
        );

        logger.info("Created Limbo Server: {}, on port {}", limboServer.getConfig().getBrandName(), port);
        return limboServer;
    }

    public void destroyLimboServer(LimboServer limboServer) {
        int port = limboServers.get(limboServer);
        limboServer.stop();
        limboServers.remove(limboServer);
        logger.info("Destroyed Limbo Server: {}, on port {}", limboServer.getConfig().getBrandName(), port);
    }

    private static class Config implements LimboConfig {
        private final String name;
        private final int port;
        private final InfoForwarding infoForwarding;
        private final PingData pingData;

        public Config(String name, int port) {
            this.name = name;
            this.port = port;
            this.infoForwarding = new InfoForwarding();
            this.pingData = new PingData();
            this.pingData.setDescription("NexusLimbo");
            this.pingData.setProtocol(-1);
            this.pingData.setVersion("1.0");
        }

        @Override
        public SocketAddress getAddress() {
            return new InetSocketAddress("0.0.0.0", port);
        }

        @Override
        public int getMaxPlayers() {
            return -1;
        }

        @Override
        public PingData getPingData() {
            return pingData;
        }

        @Override
        public String getDimensionType() {
            return "THE_END";
        }

        @Override
        public int getGameMode() {
            return 3;
        }

        @Override
        public InfoForwarding getInfoForwarding() {
            return infoForwarding;
        }

        @Override
        public long getReadTimeout() {
            return 30000;
        }

        @Override
        public int getDebugLevel() {
            return 0;
        }

        @Override
        public boolean isUseBrandName() {
            return true;
        }

        @Override
        public boolean isUseJoinMessage() {
            return false;
        }

        @Override
        public boolean isUseBossBar() {
            return false;
        }

        @Override
        public boolean isUseTitle() {
            return false;
        }

        @Override
        public boolean isUsePlayerList() {
            return false;
        }

        @Override
        public boolean isUseHeaderAndFooter() {
            return false;
        }

        @Override
        public String getBrandName() {
            return name;
        }

        @Override
        public String getJoinMessage() {
            return "";
        }

        @Override
        public BossBar getBossBar() {
            return null;
        }

        @Override
        public Title getTitle() {
            return null;
        }

        @Override
        public String getPlayerListUsername() {
            return "";
        }

        @Override
        public String getPlayerListHeader() {
            return "";
        }

        @Override
        public String getPlayerListFooter() {
            return "";
        }

        @Override
        public boolean isUseEpoll() {
            return true;
        }

        @Override
        public int getBossGroupSize() {
            return 1;
        }

        @Override
        public int getWorkerGroupSize() {
            return 4;
        }
    }


    private static class CommandHandler_ implements CommandHandler<Command> {
        @Override
        public Collection<Command> getCommands() {
            return List.of();
        }

        @Override
        public void register(Command command) {
        }

        @Override
        public boolean executeCommand(String s) {
            return false;
        }
    }
}
`

`
public void onLimboRequest(Player player) {
        LimboServer server = plugin.getLimboManager().createLimboServer();
        if (server == null) {
            player.disconnect(Component.text("Unable to create limbo server! (Very Bad)"));
            return;
        }

        server.getConfig().getPingData().setProtocol(player.getProtocolVersion().getProtocol());
        server.getConfig().getPingData().setVersion(player.getProtocolVersion().getVersionIntroducedIn());

        player.createConnectionRequest(proxyServer.getServer(server.getConfig().getBrandName()).get()).fireAndForget();
    }
    `
    
    `
    [17:55:57 ERROR]: [connected player] Milkthedev (ip): disconnected while connecting to local-limbo-19257: Unsupported client version
    `
